src/input/input_python.rs

```diff
@@ -240,12 +242,34 @@ impl<'a> Input<'a> for PyAny {
             // Safety: the gil is held while from_utf8 is running so py_byte_array is not mutated,
             // and we immediately copy the bytes into a new Python string
             let s = match from_utf8(unsafe { py_byte_array.as_bytes() }) {
-                // Why Python not Rust? to avoid an unnecessary allocation on the Rust side, the
+                // Why Python not Rust? to avoid an unnecessary allocation o(n the Rust side, the
                 // final output needs to be Python anyway.
                 Ok(s) => PyString::new(self.py(), s),
                 Err(_) => return Err(ValError::new(ErrorTypeDefaults::StringUnicode, self)),
             };
             Ok(s.into())
+        } else if coerce_numbers_to_str {
+            if let Ok(py_decimal) = self.lax_decimal(py) {
```

Comment:
The downside of going through `lax_decimal` will be performance - you're creating a whole new Python `decimal` on the way.

Reply:
Yes, it is. We knew it will be less performant feature anyway.

It hits this only if it's not a string which is already the "bad" case and the alternative is hitting several TypeError exceptions. The alternative is to almost copy the `lax_decimal` code.

I'll give it another try.

Reply:
Implemented using just `isinstance` calls.